### PR TITLE
Viewport rotation support in composer

### PR DESCRIPTION
**Pull request by strk**
This is a work in progress, see 
http://hub.qgis.org/issues/11912

### CODE REVIEW

**Comment by nyalldawson** [comment]
@strk just a note - the code in QgsComposerMap is extremely fragile, so please don't merge any changes before I've had a chance to review them

**Comment by strk** [comment]
@nyalldawson I don't plan to merge this before a deep review and an approval.
With latest commit the position of a rotated map seems fine to me, could you please take a look ?
\cc @nirvn

**Comment by nirvn** [comment]
@strk @nyalldawson , I've built and tested this PR and confirm that the rotation works fine, as well as grid and overview who now align flawlessly:

working

IMO, this should be merged before the release of 2.8; I've always considered the lack of implementation within the composer as an implementation gap (which is considered a blocker).

**Comment by gioman** [comment]
If you plan to merge this do it as soon as possible, to allow us do extensive tests.

**Comment by strk** [comment]
Travis is failing, I guess some tests would need to be changed to expect the now-correct behavior.
Also, I'd like a word from the release manager about putting this in master. \cc @jef-n

**Comment by strk** [comment]
The travis failures are here: http://dash.orfeo-toolbox.org/testDetails.php?test=28184445&build=170932

That test is particularly unfortunate in that it's using a pretty noisy raster which makes comparing expected/obtained harder than it could be. Ideally the raster would be a neatly defined 2-colors kind of thing...

**Comment by nirvn** [comment]
I gave it further testing this afternoon (testing it out in the context of an atlas export, adding an overview onto a rotated map, etc.), things are looking good. Used two large projects too, couldn't see any issues.

**Comment by strk** [comment]
If anyone can help with the failing testcases my branch is open for pull requests - https://github.com/strk/qgis/pulls

**Comment by strk** [comment]
See #1884 for the direction I have in mind for the tests...

**Comment by strk** [comment]
I've rebased to master after simplifying the test raster. Hopefully now we can see better what fails and how different the results are from the expected (debatable expectances when it comes to center, btw)

**Comment by strk** [comment]
Great, failures seem tolerable:
http://dash.orfeo-toolbox.org/testDetails.php?test=28205994&build=171052
It may be better to use a tolerance this time, rather than an "anomaly"

As for the second failure, I think it may be a completely unrelated race condition (presence of the old image in the second test): http://dash.orfeo-toolbox.org/testDetails.php?test=28205995&build=171052

**Comment by strk** [comment]
See ? now the second test did not fail. Maybe the QVERIFY failure avoids to run the subsequent code, which includes a "removeItem"...

In any case, this PR is now succeeding. Anything preventing a merge to master ?

**Comment by nyalldawson** [comment]
@strk shouldn't be. Give me 48 hours or so and I'll review and merge. Thanks!

**Comment by wonder-sk** [comment]
@strk I think it would be very useful to move the code that calculates polygon of the rotated extent to QgsMapSettings class - the visibleExtent() method only returns bounding box of the rotated extent...

**Comment by strk** [comment]
@wonder-sk I think I'm going to do that as part of the change that fixes the overview map.
Here I think the work can be completed by removing warnings and squashing to master, which I'll be doing now.

**Comment by strk** [comment]
@wonder-sk PR #1888 has the QgsMapSettings method to get possibly rotated map extent, being a new interface I dunno if it's ok to push to master (although completely safe).

**Comment by strk** [comment]
I've squashed the composer commits and rebased on master + visiblePolygon PR (#1888).

**Comment by strk** [comment]
Sorry, rebased again, the "visiblePolygon" interface is not really needed for this PR and I don't want to mix things. Better to merge this separately from that, as long as all tests still succeed.

File: src/core/composer/qgscomposermap.cpp [hunk 1]
```diff
@@ -2055,7 +2032,7 @@ QString QgsComposerMap::displayName() const
 void QgsComposerMap::requestedExtent( QgsRectangle& extent ) const
 {
   QgsRectangle newExtent = *currentMapExtent();
-  if ( mEvaluatedMapRotation == 0 )
+  if ( 1 || mEvaluatedMapRotation == 0 )
```
Comment (nyalldawson) [inline src/core/composer/qgscomposermap.cpp]: @strk I don't think this change should be included. It breaks API and affects QgsComposerLegend::updateFilterByMap() potentially causing features to be missed in a legend filtered from a rotated map (although the current behaviour isn't ideal as it potentially includes features which aren't visible, but that's safer then the alternative). Instead, the calls to requestedExtent in cache and paint should be replaced by *currentMapExtent() instead.

**Comment by nyalldawson** [comment]
@strk looks good to me, apart from the note about QgsComposerMap::requestedExtent breakage. I'm happy for it to be merged otherwise.

Small tip - including "fix #11912" in the commit message will auto-close the bug in redmine.

**Comment by strk** [comment]
@nyalldawson could you add a testcase guarding the API for QgsComposerMap::requestedExtent ?
Or suggest one ? I'll work on reverting that change and changing the callers instead.

**Comment by strk** [comment]
There are multiple callers but I can't find a way to trigger all of them. For example, how do I set "plotStyle" to "preview" ?

**Comment by nyalldawson** [comment]
@strk it's set by the composition - a composition is in preview mode in the GUI, and print/postscript if it's bring exported.

**Comment by strk** [comment]
I've pushed a fix of the regression you pointed out. Please test again.
Once happy I'll squash-rebase  and add the `fix  #11912` label

**Comment by strk** [comment]
Pushed as a40eca4a343eacff8c610e7f945e675a0edf0ec6

**Pull request closed without merging**